pkg/handlers/default.go

```diff
@@ -0,0 +1,245 @@
+func (h *DefaultHandler) openArchive(ctx logContext.Context, depth int, reader io.Reader, archiveChan chan []byte) error {
+	if common.IsDone(ctx) {
+		return ctx.Err()
+	}
+
+	if depth > maxDepth {
+		return ErrMaxDepthReached
+	}
+
+	format, arReader, err := archiver.Identify("", reader)
+	switch {
+	case err == nil:
+		// Continue with the rest of the code.
+	case errors.Is(err, archiver.ErrNoMatch):
+		if depth > 0 {
+			// If openArchive is called on an already extracted/decompressed file and the depth is greater than 0,
+			// it means we are at least 1 layer deep in the archive. In this case, we should handle the content
+			// as non-archive data by calling handleNonArchiveContent.
+			return h.handleNonArchiveContent(ctx, arReader, archiveChan)
+		}
+		// If openArchive is called on the root (depth == 0) and we can't identify the format,
+		// it means we can't handle the content at all. Return the archiver.ErrNoMatch error.
+		return err
+	default:
+		// Some other error occurred.
+		return fmt.Errorf("error identifying archive: %w", err)
+	}
+
+	switch archive := format.(type) {
+	case archiver.Decompressor:
+		// Decompress tha archive and feed the decompressed data back into the archive handler to extract any nested archives.
+		compReader, err := archive.OpenReader(arReader)
+		if err != nil {
+			return err
+		}
+
+		defer compReader.Close()
+
+		return h.openArchive(ctx, depth+1, compReader, archiveChan)
+	case archiver.Extractor:
+		return archive.Extract(logContext.WithValue(ctx, depthKey, depth+1), arReader, nil, h.extractorHandler(archiveChan))
+	default:
+		return fmt.Errorf("unknown archive type: %s", format.Name())
+	}
+}
+
+// extractorHandler creates a closure that handles individual files extracted by an archiver.
+// It logs the extraction, checks for cancellation, and decides whether to skip the file based on its name or type,
+// particularly for binary files if configured to skip. If the file is not skipped, it recursively calls openArchive
+// to handle nested archives or to continue processing based on the file's content and depth in the archive structure.
+func (h *DefaultHandler) extractorHandler(archiveChan chan []byte) func(context.Context, archiver.File) error {
+	return func(ctx context.Context, file archiver.File) error {
+		lCtx := logContext.WithValues(
+			logContext.AddLogger(ctx),
+			"filename", file.Name(),
+			"size", file.Size(),
+		)
+		lCtx.Logger().V(5).Info("Handling extracted file.")
+
+		if common.IsDone(ctx) {
+			return ctx.Err()
+		}
+
+		depth := 0
+		if ctxDepth, ok := ctx.Value(depthKey).(int); ok {
+			depth = ctxDepth
+		}
+		if int(file.Size()) > maxSize {
+			lCtx.Logger().V(3).Info("skipping file due to size")
+			return nil
+		}
+
+		if common.SkipFile(file.Name()) {
+			lCtx.Logger().V(5).Info("skipping file")
+			return nil
+		}
+
+		if h.skipBinaries && common.IsBinary(file.Name()) {
```

Comment:
there are two "skip binaries" checks in this pr - do they get unified? later in your change stack?

Reply:
Yea it should eventually get consolidated into a single check within the defaultHandler.